File: swift-bridge/Sources/SmixSDK/HttpSmixSimRuntime.swift
// HttpSmixSimRuntime — v1.0.3.
//
// URLSession-backed `SmixSimRuntime` speaking the SmixRunnerCore HTTP
// wire (mirror of TS `HttpSimRuntime` and Rust `HttpRunnerClient`).
// Adds session lifecycle: when `sessionId` is set via `setSessionId(_:)`,
// every subsequent request carries the `Session-Id` header so the
// runner reuses the session's cached `XCUIApplication` binding and
// skips per-request `.activate()`.
//
// Consumers who need session lifecycle use the `Session` type
// (Session.swift) which drives `/session/open|close|renew-activation`
// through this runtime. Consumers who want the legacy per-request path
// simply skip Session; every request goes out without `Session-Id`.

import Foundation
#if canImport(CoreGraphics)
import CoreGraphics
#endif

/// URLSession-backed runtime targeting a running SmixRunnerCore
/// instance at a client-supplied base URL. Session-aware.
///
/// Not thread-safe on `sessionId` mutation — call `setSessionId(_:)`
/// only from the same actor / dispatch queue that issues the requests.
public final class HttpSmixSimRuntime: SmixSimRuntime, @unchecked Sendable {
    public let baseURL: URL
    public let bundleId: String
    private let urlSession: URLSession
    private var sessionId: String?

    public init(
        baseURL: URL,
        bundleId: String,
        urlSession: URLSession = .shared
    ) {
        self.baseURL = baseURL
        self.bundleId = bundleId
        self.urlSession = urlSession
    }

    /// v1.0.3 — attach / clear the `Session-Id` header on every
    /// subsequent request. Typically driven by `Session.open` /
    /// `Session.close`. Consumers can call directly if they manage
    /// sessions manually.
    public func setSessionId(_ id: String?) {
        self.sessionId = id
    }

    /// The current session id, or nil when no session is attached.
    public var currentSessionId: String? {
        sessionId
    }

    // MARK: - SmixSimRuntime protocol

    public func launch(bundleId: String) async throws {
        try await postVoid("/sim/launch", ["bundleId": bundleId])
    }

    public func terminate(bundleId: String) async throws {
        try await postVoid("/sim/terminate", ["bundleId": bundleId])
    }

    public func snapshotTree() async throws -> A11yNode {
        struct Wrap: Decodable { let tree: A11yNode }
        let w: Wrap = try await post("/a11y/snapshot", [String: String]())
        return w.tree
    }

    public func synthesizeTap(at point: CGPoint) async throws {
        try await postVoid(
            "/input/tap",
            ["x": Double(point.x), "y": Double(point.y)]
        )
    }

    public func sendString(_ text: String) async throws {
        try await postVoid("/input/send-string", ["text": text])
    }

    public func pressKey(_ key: KeyName) async throws {
        try await postVoid("/input/press-key", ["key": key.wireName])
    }

    public func swipe(_ direction: SwipeDirection) async throws {
        try await postVoid("/input/swipe", ["direction": direction.wireName])
    }

    public func screenshot() async throws -> Data {
        struct Wrap: Decodable { let png: String }
        let w: Wrap = try await post("/sim/screenshot", [String: String]())
        guard let bytes = Data(base64Encoded: w.png) else {
            throw HttpSmixSimRuntimeError.malformedResponse(
                path: "/sim/screenshot",
                detail: "png base64 decode failed"
            )
        }
        return bytes
    }

    public func systemPopups() async throws -> [A11yNode] {
        struct Wrap: Decodable { let nodes: [A11yNode] }
        let w: Wrap = try await post("/sim/system-popups", [String: String]())
        return w.nodes
    }

    public func openUrl(_ url: URL) async throws {
        try await postVoid("/sim/open-url", ["url": url.absoluteString])
    }

    public func launchFresh(
        bundleId: String,
        clearState: Bool,
        clearKeychain: Bool,
        appPath: String?
    ) async throws {
        var body: [String: Any] = [
            "bundleId": bundleId,
            "clearState": clearState,
            "clearKeychain": clearKeychain,
        ]
        if let p = appPath {
            body["appPath"] = p
        }
        try await postVoid("/sim/launch-fresh", body)
    }

    public func launchFromPath(_ appPath: String) async throws {
        try await postVoid("/sim/launch-from-path", ["appPath": appPath])
    }

    public func synthesizeTapAtNormalized(nx: Double, ny: Double) async throws {
        try await postVoid("/input/tap-normalized", ["nx": nx, "ny": ny])
    }

    // MARK: - internal HTTP helpers

    /// POST that expects a JSON body back. Errors on non-2xx or decode
    /// failure.
    func post<Body: Encodable, Resp: Decodable>(
        _ path: String,
        _ body: Body
    ) async throws -> Resp {
        let data = try await requestData(path, body: body)
        do {
            return try JSONDecoder().decode(Resp.self, from: data)
        } catch {
            throw HttpSmixSimRuntimeError.malformedResponse(
                path: path,
                detail: "decode \(Resp.self): \(error)"
            )
        }
    }

    /// POST when body encoding uses `[String: Any]` (JSONSerialization
    /// path). Same shape as the strongly-typed `post`.
    func post<Resp: Decodable>(
        _ path: String,
        _ body: [String: Any]
    ) async throws -> Resp {
        let data = try await requestData(path, jsonObject: body)
        do {
            return try JSONDecoder().decode(Resp.self, from: data)
        } catch {
            throw HttpSmixSimRuntimeError.malformedResponse(
                path: path,
                detail: "decode \(Resp.self): \(error)"
            )
        }
    }

    /// POST that ignores the response body (used for void endpoints
    /// like `/sim/launch`). Errors on non-2xx.
    func postVoid<Body: Encodable>(_ path: String, _ body: Body) async throws {
        _ = try await requestData(path, body: body)
    }

    /// POST with `[String: Any]` body ignoring the response.
    func postVoid(_ path: String, _ body: [String: Any]) async throws {
        _ = try await requestData(path, jsonObject: body)
    }

    private func requestData<Body: Encodable>(
        _ path: String,
        body: Body
    ) async throws -> Data {
        let payload: Data
        do {
            payload = try JSONEncoder().encode(body)
        } catch {
            throw HttpSmixSimRuntimeError.encodeFailed(
                path: path,
                detail: "\(error)"
            )
        }
        return try await requestData(path, payload: payload)
    }

    private func requestData(
        _ path: String,
        jsonObject: [String: Any]
    ) async throws -> Data {
        let payload: Data
        do {
            payload = try JSONSerialization.data(withJSONObject: jsonObject)
        } catch {
            throw HttpSmixSimRuntimeError.encodeFailed(
                path: path,
                detail: "\(error)"
            )
        }
        return try await requestData(path, payload: payload)
    }

    /// Actual URLSession dispatch. Attaches `Session-Id` header when
    /// the runtime has one attached.
    private func requestData(
        _ path: String,
        payload: Data
    ) async throws -> Data {
        var request = URLRequest(url: baseURL.appendingPathComponent(String(path.dropFirst())))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(bundleId, forHTTPHeaderField: "App-Bundle-Id")
        if let sid = sessionId {
            request.setValue(sid, forHTTPHeaderField: "Session-Id")
        }
        request.httpBody = payload
        let (data, response) = try await urlSession.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw HttpSmixSimRuntimeError.nonHttpResponse(path: path)
        }
        guard (200 ..< 300).contains(http.statusCode) else {
            let body = String(data: data, encoding: .utf8) ?? ""
            throw HttpSmixSimRuntimeError.nonSuccessStatus(
                path: path,
                status: http.statusCode,
                body: body
            )
        }
        return data
    }
}

/// Errors thrown by `HttpSmixSimRuntime`.
public enum HttpSmixSimRuntimeError: Error, LocalizedError {
    case encodeFailed(path: String, detail: String)
    case nonHttpResponse(path: String)
    case nonSuccessStatus(path: String, status: Int, body: String)
    case malformedResponse(path: String, detail: String)

    public var errorDescription: String? {
        switch self {
        case let .encodeFailed(path, detail):
            return "smix runner \(path): encode failed: \(detail)"
        case let .nonHttpResponse(path):
            return "smix runner \(path): non-HTTP response"
        case let .nonSuccessStatus(path, status, body):
            return "smix runner \(path): HTTP \(status): \(body)"
        case let .malformedResponse(path, detail):
            return "smix runner \(path): malformed body: \(detail)"
        }
    }
}

// MARK: - key + direction wire names

extension KeyName {
    /// Wire string used by `POST /input/press-key {key}` — mirrors
    /// the Rust `KeyName` variant name in the smix-runner-wire crate.
    public var wireName: String {
        switch self {
        case .`return`: return "Return"
        case .delete: return "Delete"
        case .space: return "Space"
        case .tab: return "Tab"
        case .escape: return "Escape"
        case .enter: return "Enter"
        }
    }
}

extension SwipeDirection {
    /// Wire string used by `POST /input/swipe {direction}` — mirrors
    /// the Rust `SwipeDirection` variant name.
    public var wireName: String {
        rawValue  // enum's rawValue is already "up" / "down" / "left" / "right"
    }
}

File: android-runner/sdk/src/main/kotlin/dev/smix/sdk/HttpSmixSimRuntime.kt
// v1.0.3 — HttpSmixSimRuntime + Session wiring for Kotlin SDK.
//
// java.net.HttpURLConnection–backed SmixSimRuntime implementation that
// speaks the SmixRunnerCore wire (mirror of TS HttpSimRuntime and
// Swift HttpSmixSimRuntime). Adds session lifecycle: when a session id
// is attached via setSessionId(id), every subsequent request carries
// the Session-Id header so the runner reuses the session's cached
// UiAutomator handle and skips per-request .activate() equivalents.
//
// Consumers who need session lifecycle use the Session class
// (Session.kt) which drives /session/open|close|renew-activation via
// this runtime. Consumers who want the legacy per-request path skip
// Session entirely; every request goes out without Session-Id.

package dev.smix.sdk

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.serializer
import java.net.HttpURLConnection
import java.net.URL
import java.util.Base64
import java.util.concurrent.atomic.AtomicReference

/**
 * HTTP-backed [SmixSimRuntime] targeting a running SmixRunnerCore
 * instance at a client-supplied base URL.
 *
 * Session-aware: when a session id is attached via [setSessionId],
 * every request carries the `Session-Id` header and the runner reuses
 * the session's cached binding.
 *
 * Thread-safe on the session-id field (AtomicReference), so consumers
 * may share a single runtime across coroutine dispatchers.
 */
class HttpSmixSimRuntime(
    val baseUrl: String,
    val bundleId: String,
    private val connectTimeoutMs: Int = 5_000,
    private val readTimeoutMs: Int = 15_000,
) : SmixSimRuntime {

    private val sessionIdRef = AtomicReference<String?>(null)

    /** JSON codec — lenient so pre-v1.0.3 runners without new fields still decode. */
    val json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = true
    }

    /**
     * v1.0.3 — attach or clear the `Session-Id` header on every
     * subsequent request. Typically driven by [Session.open] /
     * [Session.close]; consumers who manage sessions manually may call
     * directly.
     */
    fun setSessionId(id: String?) {
        sessionIdRef.set(id)
    }

    /** Current session id, or null when no session is attached. */
    val currentSessionId: String?
        get() = sessionIdRef.get()

    // ---- SmixSimRuntime protocol ------------------------------------

    override suspend fun launch(bundleId: String) {
        postVoid("/sim/launch", JsonObject(mapOf("bundleId" to JsonPrimitive(bundleId))))
    }

    override suspend fun terminate(bundleId: String) {
        postVoid("/sim/terminate", JsonObject(mapOf("bundleId" to JsonPrimitive(bundleId))))
    }

    override suspend fun snapshotTree(): A11yNode {
        val bytes = postRaw("/a11y/snapshot", JsonObject(emptyMap()))
        val obj = json.parseToJsonElement(bytes.decodeToString()).jsonObject
        val treeElem = obj["tree"] ?: error("smix runner /a11y/snapshot: missing tree field")
        return json.decodeFromJsonElement(serializer(), treeElem)
    }

    override suspend fun synthesizeTap(x: Double, y: Double) {
        postVoid("/input/tap", JsonObject(mapOf(
            "x" to JsonPrimitive(x),
            "y" to JsonPrimitive(y),
        )))
    }

    override suspend fun sendString(text: String) {
        postVoid("/input/send-string", JsonObject(mapOf("text" to JsonPrimitive(text))))
    }

    override suspend fun pressKey(key: KeyName) {
        postVoid("/input/press-key", JsonObject(mapOf("key" to JsonPrimitive(key.wireName))))
    }

    override suspend fun swipe(direction: SwipeDirection) {
        postVoid("/input/swipe", JsonObject(mapOf("direction" to JsonPrimitive(direction.wireName))))
    }

    override suspend fun screenshot(): ByteArray {
        val bytes = postRaw("/sim/screenshot", JsonObject(emptyMap()))
        val obj = json.parseToJsonElement(bytes.decodeToString()).jsonObject
        val b64 = obj["png"]?.jsonPrimitive?.content
            ?: error("smix runner /sim/screenshot: missing png field")
        return Base64.getDecoder().decode(b64)
    }

    override suspend fun systemPopups(): List<A11yNode> {
        val bytes = postRaw("/sim/system-popups", JsonObject(emptyMap()))
        val obj = json.parseToJsonElement(bytes.decodeToString()).jsonObject
        val nodes = obj["nodes"] ?: return emptyList()
        return json.decodeFromJsonElement(serializer(), nodes)
    }

    override suspend fun openUrl(url: String) {
        postVoid("/sim/open-url", JsonObject(mapOf("url" to JsonPrimitive(url))))
    }

    override suspend fun launchFresh(
        bundleId: String,
        clearState: Boolean,
        clearKeychain: Boolean,
        appPath: String?,
    ) {
        val entries = mutableMapOf<String, JsonElement>(
            "bundleId" to JsonPrimitive(bundleId),
            "clearState" to JsonPrimitive(clearState),
            "clearKeychain" to JsonPrimitive(clearKeychain),
        )
        if (appPath != null) {
            entries["appPath"] = JsonPrimitive(appPath)
        }
        postVoid("/sim/launch-fresh", JsonObject(entries))
    }

    override suspend fun launchFromPath(appPath: String) {
        postVoid("/sim/launch-from-path", JsonObject(mapOf("appPath" to JsonPrimitive(appPath))))
    }

    override suspend fun synthesizeTapAtNormalized(nx: Double, ny: Double) {
        postVoid("/input/tap-normalized", JsonObject(mapOf(
            "nx" to JsonPrimitive(nx),
            "ny" to JsonPrimitive(ny),
        )))
    }

    // ---- internal HTTP helpers --------------------------------------

    /**
     * v1.0.3 — used by [Session] to drive the `/session/open`,
     * `/session/close`, and `/session/renew-activation` routes through
     * the same transport.
     */
    internal fun postJsonObject(path: String, body: JsonObject): JsonObject {
        val bytes = postRaw(path, body)
        return json.parseToJsonElement(bytes.decodeToString()).jsonObject
    }

    /** POST expecting no meaningful body. */
    internal fun postVoid(path: String, body: JsonElement) {
        postRaw(path, body)
    }

    /** POST returning raw bytes. Throws on non-2xx. */
    internal fun postRaw(path: String, body: JsonElement): ByteArray {
        val url = URL(baseUrl.trimEnd('/') + path)
        val conn = url.openConnection() as HttpURLConnection
        try {
            conn.connectTimeout = connectTimeoutMs
            conn.readTimeout = readTimeoutMs
            conn.requestMethod = "POST"
            conn.doOutput = true
            conn.setRequestProperty("Content-Type", "application/json")
            conn.setRequestProperty("App-Bundle-Id", bundleId)
            val sid = sessionIdRef.get()
            if (sid != null) {
                conn.setRequestProperty("Session-Id", sid)
            }
            val payload = json.encodeToString(JsonElement.serializer(), body).toByteArray()
            conn.outputStream.use { it.write(payload) }
            val status = conn.responseCode
            val stream = if (status in 200..299) conn.inputStream else conn.errorStream
            val bytes = stream?.readBytes() ?: ByteArray(0)
            if (status !in 200..299) {
                throw HttpSmixSimRuntimeException(
                    "smix runner $path: HTTP $status: ${String(bytes)}"
                )
            }
            return bytes
        } finally {
            conn.disconnect()
        }
    }
}

/** Errors thrown by [HttpSmixSimRuntime]. */
class HttpSmixSimRuntimeException(message: String) : RuntimeException(message)

// ---- wire name mappings ---------------------------------------------

val KeyName.wireName: String
    get() = when (this) {
        KeyName.RETURN -> "Return"
        KeyName.DELETE -> "Delete"
        KeyName.SPACE -> "Space"
        KeyName.TAB -> "Tab"
        KeyName.ESCAPE -> "Escape"
        KeyName.ENTER -> "Enter"
    }

val SwipeDirection.wireName: String
    get() = when (this) {
        SwipeDirection.UP -> "up"
        SwipeDirection.DOWN -> "down"
        SwipeDirection.LEFT -> "left"
        SwipeDirection.RIGHT -> "right"
    }

